translate/v33tov22/v33tov22.go

```diff
@@ -0,0 +1,470 @@
+
+	old "github.com/coreos/ignition/config/v2_2/types"
+	oldValidate "github.com/coreos/ignition/config/validate"
+	"github.com/coreos/ignition/v2/config/v3_3/types"
+	"github.com/coreos/ignition/v2/config/validate"
+
+	"github.com/coreos/ign-converter/util"
+)
+
+// Translate translates Ignition spec config v3.3 to v2.2
+func Translate(cfg types.Config) (old.Config, error) {
+	rpt := validate.ValidateWithContext(cfg, nil)
+	if rpt.IsFatal() {
+		return old.Config{}, fmt.Errorf("Invalid input config:\n%s", rpt.String())
+	}
+
+	// Check for potential issues in the spec 3 config
+	for _, m := range cfg.Ignition.Config.Merge {
+		if m.Compression != nil {
+			return old.Config{}, fmt.Errorf("Compression in Ignition.Config.Merge is not supported on 2.2")
+		}
+		if m.HTTPHeaders != nil {
+			return old.Config{}, fmt.Errorf("HTTPHeaders in Ignition.Config.Merge are not supported on 2.2")
+		}
+	}
+
+	if cfg.Ignition.Config.Replace.Compression != nil {
+		return old.Config{}, fmt.Errorf("Compression in Ignition.Config.Replace is not supported on 2.2")
+	}
+
+	if cfg.Ignition.Config.Replace.HTTPHeaders != nil {
+		return old.Config{}, fmt.Errorf("HTTPHeaders in Ignition.Config.Replace are not supported on 2.2")
+	}
+
+	for _, ca := range cfg.Ignition.Security.TLS.CertificateAuthorities {
+		if ca.Compression != nil {
+			return old.Config{}, fmt.Errorf("Compression in Ignition.Security.TLS.CertificateAuthorities is not supported on 2.2")
+		}
+		if ca.HTTPHeaders != nil {
+			return old.Config{}, fmt.Errorf("HTTPHeaders in Ignition.Security.TLS.CertificateAuthorities are not supported on 2.2")
+		}
+	}
+
+	if cfg.Ignition.Proxy.HTTPProxy != nil || cfg.Ignition.Proxy.HTTPSProxy != nil || cfg.Ignition.Proxy.NoProxy != nil {
+		return old.Config{}, fmt.Errorf("HTTP proxies in Ignition.Proxy are not supported on 2.2")
+	}
+
+	if len(cfg.Storage.Luks) > 0 {
+		return old.Config{}, fmt.Errorf("LUKS is not supported on 2.2")
+	}
+
+	// ShouldExist for Users & Groups do not exist in 2.2
+	for _, u := range cfg.Passwd.Users {
+		if u.ShouldExist != nil && !*u.ShouldExist {
+			return old.Config{}, fmt.Errorf("ShouldExist in Passwd.Users is not supported on 2.2")
+		}
+	}
+	for _, g := range cfg.Passwd.Groups {
+		if g.ShouldExist != nil && !*g.ShouldExist {
+			return old.Config{}, fmt.Errorf("ShouldExist in Passwd.Groups is not supported on 2.2")
+		}
+	}
+
+	// Size and Start are sectors not MiB in 2.2, so we don't understand them.
+	// Resize is not in 2.2
+	// Fail for now
+	for _, d := range cfg.Storage.Disks {
+		for _, p := range d.Partitions {
+			if p.SizeMiB != nil || p.StartMiB != nil {
+				return old.Config{}, fmt.Errorf("SizeMiB and StartMiB in Storage.Disks.Partitions is not supported on 2.2")
+			}
+			if p.Resize != nil && *p.Resize {
+				return old.Config{}, fmt.Errorf("Resize in Storage.Disks.Partitions is not supported on 2.2")
+			}
+		}
+	}
+
+	if len(cfg.KernelArguments.ShouldExist) > 0 || len(cfg.KernelArguments.ShouldNotExist) > 0 {
```

Comment:
I understand how this ended up here, but let's move it down two stanzas, after the other `cfg.Storage` checks.